lib/elsa/producer.ex

```diff
@@ -2,22 +2,57 @@ defmodule Elsa.Producer do
   @moduledoc """
   Defines functions to send messages to topics based on either a list of endpoints or a named client.
   """
-  def produce_sync(client \\ Elsa.default_client(), topic, partition \\ 0, key \\ "ignored", value)
 
-  def produce_sync(endpoints, topic, partition, key, value) when is_list(endpoints) do
-    client = Elsa.default_client()
+  alias Elsa.Util
+
+  def produce(endpoints, topic, messages, opts \\ []) when is_list(endpoints) do
+    client = get_client(opts)
 
     Elsa.Producer.Manager.start_producer(endpoints, topic, name: client)
-    produce_sync(client, topic, partition, key, value)
+    produce_sync(topic, messages, Keyword.put(opts, :client, client))
+  end
+
+  def produce_sync(topic, messages, opts \\ [])
+
+  def produce_sync(topic, messages, opts) when is_list(messages) do
+    client = get_client(opts)
+
+    case Keyword.get(opts, :partition) do
+      nil ->
+        messages
+        |> Enum.map(fn {key, value} -> {get_partition(client, topic, key, opts), {key, value}} end)
+        |> Enum.group_by(fn {partition, _} -> partition end, fn {_, message} -> message end)
```

Comment:
on the one hand i like this; on the other hand, does it feel like this is starting to step into broadway territory?

Reply:
Broadway is focused on consuming, so this would fit well with Broadway from what I can remember.

Broadway consumes off something, batches the messages, and passes a batch of messages to Elsa to produce to the topic.

Reply:
fair enough; i didn't remember it was only focused on consuming. ignore my comments.